000-vip-init.php

```diff
@@ -198,6 +198,41 @@
 	}
 }
 
+// Set WordPress >5.5 environment name
+if ( defined( 'VIP_GO_APP_ENVIRONMENT' ) && ! defined( 'WP_ENVIRONMENT_TYPE' ) ) {
+	$environment_name = VIP_GO_APP_ENVIRONMENT;
+
+	// Adapt some VIP environment names to the WordPress standards
+	switch ( $environment_name ) {
+		case 'develop':
+			$environment_name = 'development';
+			break;
+		case 'stage':
+			$environment_name = 'staging';
+			break;
+	}
+
+	define( 'WP_ENVIRONMENT_TYPE', $environment_name );
+
+	if ( ! defined( 'WP_ENVIRONMENT_TYPES' ) ) {
+		// Basic set of VIP environments (WP ones plus 'testing' and 'preprod')
+		$environment_types = array(
+			'development',
+			'testing',
+			'staging',
+			'preprod',
+			'production',
+		);
+
+		// Add the environment to the environment type list if it's a non standard name
+		if ( ! in_array( $environment_name, $environment_types ) ) {
```

Comment:
`in_array()` is kinda slow in PHP since it traverses the array. This is a pretty short array, but given that this runs on _every_ request, I wonder if there's a more performant way to do it. Two thoughts come to mind (I haven't verified that they're faster) - either just add the env to the array, then call `array_unique()` later, or flip it to an associative array like `development => true` so each only appears once, then convert to the WP format with `array_keys()`.

Reply:
It's also unclear from https://make.wordpress.org/core/2020/07/24/new-wp_get_environment_type-function-in-wordpress-5-5/ what the plan for `WP_ENVIRONMENT_TYPES` is. The current implementation here won't take into account the available environment types of other child environments, just the current one, which _may_ (or may not) be a problem at some point.

My worry is that if we're defining this array but not including all the environment types for a given site and it's child environments, things can get confusing. As an example, say we have a production site with `develop`, `uat`, `uat2`, `testing` environments. Each of those will only know about itself, but the others are all "valid" environment types for that site. Thoughts?

Reply:
For the types constant, it's probably good enough to ship what we have without trying to account for all env types of a site, and improve it later by moving it to a constant generated in wp-config.php, if needed.

Reply:
> I wonder if there's a more performant way to do it.

Certainly, I also considered that. Checking the key of an array would be less expensive, but when you include `array_keys()` it becomes more expensive than `in_array()`. It's also possible to declare both arrays (one to search, and the other one to apply to the define), but it makes the code not very clean.

I've been doing [some tests](https://gist.github.com/netsuso/0a557e2f6f64bc3e77ffde53aeb4c41a), and in summary there are three options:

- in_array (Code 1): 15 nanoseconds
- two arrays (Code 3): : 9 nanoseconds
- if with shortcut for production (Code 5): 4 nanoseconds

Not that Code 5 is very clear, but at least looks nicer than Code 3 with its two arrays. At least it's the fastest solution for production environments (which are the most sensitive). In any case, we are talking about very short times...

Reply:
> The current implementation here won't take into account the available environment types

Agree with that, it seems somewhat incomplete... And also agree that this may be revisited in the future if needed.